### PR TITLE
[SPARK-20889][SparkR] Grouped documentation for AGGREGATE column methods

### PR DESCRIPTION
**Pull request by actuaryzhang**
## What changes were proposed in this pull request?
Grouped documentation for the aggregate functions for Column.

### CODE REVIEW

**Comment by actuaryzhang** [comment]
@felixcheung @HyukjinKwon 

Per this [suggestion](https://github.com/apache/spark/pull/18003#discussion-diff-116853922L57), I'm creating more meaningful examples for the SQL functions. 

Since these functions can be grouped, we can create a single page doc for each group of the functions and construct concrete and useful examples for each group. The benefit is obvious:
-  Centralized documentation of related functions. This makes it easier for user to navigate. Right now there are TOO many items in the `see also` section. 
-  Examples can share the same data. This avoids creating a data frame for each function if they are documented separately.
- Cleaner structure and much fewer Rd files.

Indeed, this is part of what was discussed in #17161. I have explored this for a few functions to illustrate the idea. Since this is a big effort, I would like to get folks' opinions before extending this to all functions. 

In this commit, I created docs for some sample functions in three groups:
-  'column_datetime_functions' to document all datetime functions
-  'column_aggregate_functions' to document all aggregate functions
-  'column_math_functions' to document all math functions
-  ...

Below is what 'column_datetime_functions.Rd' looks like:

image
image

**Comment by actuaryzhang** [comment]
This is what the `'column_aggregate_functions.Rd'` doc looks like:

image
image
image

**Comment by felixcheung** [comment]
so I think better example is great and we might be too verbose with individual pages for each function so might be a good idea to consolidate them, but one question, does this affect discoverability?

can function be found on http://spark.apache.org/docs/latest/api/R/index.html
also can one find the function in the shell with `?abs`?

**Comment by actuaryzhang** [comment]
@felixcheung Thanks for your feedback. 
- This does not affect discoverability: the name of the method is still on the index list 
- No problem with help either, e.g., one can use `?avg`. 

image

Another benefit is that we can get rid of most warnings on no examples since we now document all the tiny functions together. 

I think it is important and the change is straightforward. However, this is a pretty manual (and big) change. I would like to get a `Yes` from you for doing this. Thanks.

File: R/pkg/R/generics.R [hunk 1]
```diff
@@ -385,6 +385,7 @@ setGeneric("value", function(bcast) { standardGeneric("value") })
 #' @return A SparkDataFrame.
 #' @rdname summarize
 #' @export
+#' @noRd
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: what's the reason for the noRd here? I think it's fairly standard to "document" generic - even though it's not useful most of the time

File: R/pkg/R/generics.R [hunk 2]
```diff
@@ -396,67 +397,81 @@ setGeneric("agg", function (x, ...) { standardGeneric("agg") })
...
 setGeneric("cache", function(x) { standardGeneric("cache") })
 
 #' @rdname checkpoint
 #' @export
+#' @noRd
 setGeneric("checkpoint", function(x, eager = TRUE) { standardGeneric("checkpoint") })
 
 #' @rdname coalesce
 #' @param x a Column or a SparkDataFrame.
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: waiting - having these here is intentional? if a function supports different type we don't want the documentation to go to DataFrame.R because then Column.R will look empty.

Comment (felixcheung) [inline R/pkg/R/generics.R]: see here https://github.com/apache/spark/pull/18025/files#diff-04c14efaae2b7b0f0a45038482f2590cR135
how do we decide this goes to Column.R and not DataFrame.R? It's very easy then later on someone else added more comment in DataFrame.R thinking there is no documentation and then later on the help content is duplicated (has happened a few times before)

**Comment by felixcheung** [comment]
thanks for looking into this and yes it will be useful to understand better before proceeding further.

how does the `?` help search or API index page work in this case? I wonder if it's because of the `@aliases` tag (or might be `@name`?) - `aliases` we are discussing ways to get rid of since it creates duplicated links everywhere. I just want to make sure we don't break usability - people can't find help in the shell is a huge problem ;)

File: R/pkg/R/generics.R (hunk 1 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: @felixcheung I intentionally suppressed the documentation for generic functions. The main reason is the documentation of generics conflict with documenting groups of methods together. For example, if I document the `avg` method for the column class together with other column aggregate methods, here is a couple of scenarios that could happen to the generic `avg` method:
1. Without `@noRd`, roxygen will create a `avg` rd file to document the `avg` generic method ONLY. Since in the generics definition, we don't have `@title` or `@description`, the package won't compile. 
2. Suppose we get rid of `@noRd`, and document all generics functions under a new `generics` rd name. This will compile, but then we have conflicting definitions of the arguments. For example, `...` could mean different things in different generics. But this is a minor issue. 

The current approach is to suppress the documentation of the generics, since we document each method of the generics. This does not lose any information IMO. The user can always use `getGeneric("avg")` to get the prototype for the generics. Would like to hear your thoughts.

File: R/pkg/R/generics.R (hunk 2 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: Thanks for pointing this out. In this case, it should be `a Column` not `a SparkDataFrame or a Column`.  Indeed, this is a problem of documenting only the generics and inheriting the parameter doc from the generics in each method. Ideally, each specific method of a generic should have its own documentation, not the way we document the generics now.  Suppose we have a new method `alias` for another class, the doc in the generics needs to be updated to include this new class. But it's not clear to the developer where to make the change. If we instead just document each method of the generics, it will be much clearer. 
I will do a thorough cleanup of these issues after we decide on the big items.

**Comment by actuaryzhang** [comment]
@felixcheung Thanks for asking this. I should have been more clear. 

`@name` is the "name" of the Rd object represented, which is unique. An Rd object can have multiple aliases pointing to the same Rd. Usually, when one specifies `@name xxx`, there will be a `@alias xxx` automatically created to reference `xxx`.

In the `avg` case, I actually don't need `@name avg`, since the doc of `avg` goes to `column_aggregate_functions`. In the new commit, I used `@aliases avg avg,Column-method` which creates two aliases `avg` and `avg,Column-method`. 

It now allows the following shortcut for help search which all direct to the Rd object `column_aggregate_functions` (the `@name` object):

- `?avg`
- `?"avg,Column-method"`
- `help("avg")`
- `?column_aggregate_functions`
-  Indeed, since this is S4 method, we can also do `method?avg("Column")`. 

Details of R documentation here:
https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Writing-R-documentation-files

File: R/pkg/R/generics.R (hunk 1 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: hmm, I think some of the problems are orthogonal - for example, to avoid a `avg.rd` file we could just change `avg` in generic.R to have `@rdname column_aggregate_functions`?

I feel the standard is to have documentation on the generic (and have it on the same rd) like https://stat.ethz.ch/R-manual/R-devel/library/base/html/mean.html

File: R/pkg/R/generics.R (hunk 2 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: wait, I think some wires are crossed. let me clarify. let's take `coalesce` as an example.

there are 2 coalesce, one `coalesce(df)` like repartition, and one `coalesce(df$foo)` on a column like in SQL. so therefore, these are in fact 2 `coalesce` and `x is either a SparkDataFrame or a Column`.

and to elaborate, the history behind that approach we have today is because we use to have this
```
@param x a Column
...
function(x = "Column"...)
```
and at the same time, in a different .R file
```
@param x a SparkDataFrame
...
function(x = "SparkDataFrame"
```
they seem fine when we write the code and it seems logical/easy to maintain, but when the rd/doc page is generated it has
```
x a SparkDataFrame
x a Column
```
here `x` is explained twice. worse, the order is largely random (it's the alphabetic order of the .R file)

and it is going against the standard R pattern of one description line for each parameter with the choice of type separated by `.. or..` like https://stat.ethz.ch/R-manual/R-devel/library/stats/html/lm.html, and not to mention CRAN check I think complain about parameter documented more than once.

so in short, having one `@param x a SparkDataFrame or a Column` is intentional. since this is describing 2 things, from discussions back then it feels more nature to put it some where independent - in fact like you are touching on, I'd argue it's better to look it up in generic.R rather than trying to figure out what other existing overload class that method already has.

**Comment by felixcheung** [comment]
thanks, I am somewhat familiar with all the tags. I see that you have `@aliases avg` instead of `@name avg`. may I direct you to this JIRA - https://issues.apache.org/jira/browse/SPARK-18825 this is the one tracking all the double/triple links because of aliases and `@seealso`

I think we are trying to remove `@aliases avg,Column-method` (ie. `-method` ones) and so it's better to make sure we come up with a consistent way to handle this.

**Comment by actuaryzhang** [comment]
@felixcheung 
I'm curious what is the motivation to remove the style `avg,Column-method`?
This is the default and preferred (I believe) way to reference to S4 methods in R. 
Suppose we have new classes and methods on `avg`, say `avg,DataFrame-method`, `avg,RDD-method`, `avg,Row-method`, etc, is the idea to document all `avg` methods in a single doc and use one `\alias{avg}` to reference it? 

If that's the case, then the idea of grouping different methods by class will not be useful.

**Comment by felixcheung** [comment]
I guess this comes from not seeing a lot of S4 methods? In fact, a overwhelming percentage of all methods in packages are S3, and I think users are using to search with `?avg` and also expect to see different classes documented on the same page.

granted, I am aware some (many?) of these don't make sense (like coalesce(DF) and coalesce(col) have not much common) and that is a problem.

would you propose they go to different page then? would that also solve https://issues.apache.org/jira/browse/SPARK-18825 ?

**Comment by actuaryzhang** [comment]
@felixcheung I think we may want to distinguish a few cases:
1. For methods that are mainly defined by only one class, e.g., most function methods for Column, it makes sense to group and document them together. For example, most aggregate functions of Column go into one single Rd, since they are not defined for other classes. In this case, `avg` will go to this doc since it is not used by other classes. 
2. For methods that are defined by multiple classes, e.g., the `show` method defined for SparkDataFrame, GroupedData, Column and StreamingQuery, we can still document them in `show.Rd`. In this case, `show` will go to this doc and shows the help for all classes that have defined a `show` method. 
3. When it makes sense, we can also combine 1 & 2 above. For example, `gapply` and `gapplyCollecte` are defined for both SparkDataFrame and GroupedData. But we can still document them together and create shared examples. 

Let me know if this makes sense.

**Comment by felixcheung** [comment]
Thanks for summarizing. I think they make sense. To be clear though, we should also talk about:
- what if a method is defined in one class and belongs in a group, but also defined for another class (eg. sql function: `cov`)
- what if it is defined for multiple classes but meaning are drastically different (eg. coalesce(DF) and coalesce(col)  in my example above)

**Comment by actuaryzhang** [comment]
Great point. 

- For a method that is defined in one class and belongs in a group like `cov`, we can document it in its own Rd, and add a link to in the `SeeAlso` section of the group doc. In this case, the `\alias{cov}` will be in `cov.Rd`. 
- For a method that is defined for multiple classes but meaning are drastically different: I think we can still document them in one Rd, and add a `details` section to describe the method for each class.

**Comment by felixcheung** [comment]
>For a method that is defined for multiple classes but meaning are drastically different: I think we can still document them in one Rd, and add a details section to describe the method for each class.

would this be too confusing though? particularly if we are moving to -method one per doc page

**Comment by actuaryzhang** [comment]
@felixcheung I just made a new commit which I think has the cleanest solution so far. In this one, I implemented grouping for all aggregate functions for Column, except those that are also defined for other classes (`count`, `first` and `last`). As you can see, it achieves the following:
- Centralized documentation for easy navigation.
- Reduced number of items in `See also`
- Betters examples using shared data. This avoids creating a data frame for each function if they are documented separately.
- Cleaner structure and much fewer Rd files.
- Remove duplicated definition of `@param`
- No need to write meaningless examples for trivial functions (because of grouping). 

In this version, I also demonstrate the for methods defined by multiple classes (`count`, `first` and `last`), we can still document them on their own RD, and simply give a link in the `SeeAlso` section. Of course, we can combine the doc for these three to something like `shared_methods.Rd` since each of them is tiny. 

Also, to facilitate review, perhaps we can break the changes into several PRs, one for each of `aggregate_functions`, `datetime_functions`, `math_function`, and `misc_functions`? 

After making the change to the Column methods, I will work on the doc for SparkDataFrame and GroupedData. 

Please let me know your thoughts.

**Comment by actuaryzhang** [comment]
image
image
image

**Comment by felixcheung** [comment]
I think we need to give it a title explicitly - see the header/first line of https://cloud.githubusercontent.com/assets/11082368/26429381/64dd117e-409b-11e7-9661-659b5fbe8206.png

**Comment by felixcheung** [comment]
I guess we don't need link to stddev_samp since it's the same page
shouldn't std_dev and var_samp also on this page?

**Comment by felixcheung** [comment]
also, since we have an Rd now what you think about collecting all the example into one - that should eliminate all the `Not run` in every other line.

I think then also this will be a great opportunity to do more than simple `head(select(...))` something expanded and more practical? what do you think?

also this https://github.com/apache/spark/pull/18025#issuecomment-303838880

I like this approach - these are my comments from your screen shot - I'll review more closely after more changes, thanks!

**Comment by actuaryzhang** [comment]
@felixcheung 
- The links to `stddev_samp` etc are already removed in the latest commit. 
- About collecting all the example into one, I think that'll work for this particular one. But I'm not sure about this in general. These methods are still spread out in `.R` file. And if we decide to change the grouping of these functions later on, it will be very difficult if we don't have examples in those methods. 
- For a method that is defined for multiple classes but meaning are drastically different, I agree that it's best to document by class. One downside is a generic `?coalesce` can only go to one help page,  e.g., the help for SparkDataFrame, not the other classed. However, we can add links to the `coalesce` methods for the other classes in the `SeeAlso` section.

**Comment by felixcheung** [comment]
re: title, would explicitly adding `@title` help?
re: multiple class - agreed, a link or `@seealso` should be good. wouldn't `?coalesce` show the overloads though

**Comment by actuaryzhang** [comment]
- New commit now resolves the Name issue. `@title` does not work, which is the header in the second line `\title{Aggregate functions for Column operations}`. The solution is to use `@name NULL` for the generics.  Now we have:

image

- Also added several more practical examples. But most of these functions are very straightforward to use. 

image

**Comment by actuaryzhang** [comment]
@felixcheung All comments are addressed now and I think this is ready for review.

**Comment by felixcheung** [comment]
since this is a fairly big change to the documentation could you please open a JIRA? this is also how we could track this work against SPARK-18825

**Comment by actuaryzhang** [comment]
Opened a JIRA. We would need several PRs to fix all doc issues. 
Also, not sure why Jenkins failed as the error msg is not clear and all tests passed on my computer.

**Comment by felixcheung** [comment]
Jenkins, retest this please

File: R/pkg/R/generics.R [hunk 3]
```diff
@@ -1403,20 +1416,25 @@ setGeneric("unix_timestamp", function(x, format) { standardGeneric("unix_timesta
 #' @export
 setGeneric("upper", function(x) { standardGeneric("upper") })
 
-#' @rdname var
+#' @rdname column_aggregate_functions
+#' @param y,na.rm,use currently not used.
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: should this be moved like this line https://github.com/apache/spark/pull/18025/files#diff-8e3d61ff66c9ffcd6ffb7a8eedc08409L923

**Review by felixcheung — COMMENTED** [review]
I did a quick pass, will review in more details but generally looking very good.

@zero323 could you review when you get a chance?

File: R/pkg/R/generics.R (hunk 3 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: Good catch. But I  think this one makes more sense to stay here because all these arguments are specific to the `var` generic, which is not used anywhere else. I moved the `...` in `avg` to `column_aggregate_functions` to avoid duplicated doc for `...`.

**Comment by actuaryzhang** [comment]
Jenkins, retest this please

**Comment by felixcheung** [comment]
thanks! sorry about the delay - this is very important to have (and among your other pending PRs) - will pick this up again in ~ a week

**Comment by actuaryzhang** [comment]
Thanks for the update. Look forward to your feedback.

**Comment by HyukjinKwon** [comment]
Will give a shot to help double check examples and contents about correctness and consistency at my best tonight too.

**Review by HyukjinKwon — COMMENTED** [review]
I nitpicked some. Are `corr`, `covar_pop` and `covar_samp` the same instances but missed? (maybe I lost the discussion about this here).

**Comment by HyukjinKwon** [comment]
BTW, I checked the description/examples. These generally look good to me.

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -1081,19 +1098,12 @@ setMethod("md5",
...
           })
 
-#' mean
-#'
-#' Aggregate function: returns the average of the values in a group.
-#' Alias for avg.
-#'
-#' @param x Column to compute on.
+#' @section Details:
+#' \code{mean}: Returns the average of the values in a group. Alias for avg.
```
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Fixed.

File: R/pkg/R/functions.R [hunk 5]
```diff
@@ -1630,18 +1609,12 @@ setMethod("sqrt",
...
-#' @param x Column to compute on.
-#'
-#' @rdname sum
-#' @name sum
-#' @family aggregate functions
-#' @aliases sum,Column-method
+#' @rdname column_aggregate_functions
+#' @aliases sum sum,Column-method
 #' @export
-#' @examples \dontrun{sum(df$c)}
```
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Good catch. Added to example.

**Comment by actuaryzhang** [comment]
@HyukjinKwon Thanks much for the review. New commit now fixes the issues you pointed out.

File: R/pkg/R/functions.R [hunk 6]
```diff
@@ -85,17 +100,20 @@ setMethod("acos",
             column(jc)
           })
 
-#' Returns the approximate number of distinct items in a group
+#' @section Details:
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: would it be more concise to do `@details`?

File: R/pkg/R/functions.R [hunk 7]
```diff
@@ -85,17 +100,20 @@ setMethod("acos",
...
-#' @rdname approxCountDistinct
-#' @name approxCountDistinct
-#' @return the approximate number of distinct items in a group.
+#' @rdname column_aggregate_functions
 #' @export
-#' @aliases approxCountDistinct,Column-method
-#' @examples \dontrun{approxCountDistinct(df$c)}
+#' @aliases approxCountDistinct approxCountDistinct,Column-method
+#' @examples
+#'
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: extra newline?

File: R/pkg/R/functions.R [hunk 8]
```diff
@@ -2254,18 +2198,12 @@ setMethod("approxCountDistinct",
             column(jc)
           })
 
-#' Count Distinct Values
+#' @section Details:
+#' \code{countDistinct}: Returns the number of distinct items in a group.
 #'
-#' @param x Column to compute on
-#' @param ... other columns
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: this is the example perhaps we are losing some details by grouping these functions - instead of having a line that says "other columns" now we have "other arguments" which is perhaps less clear?

thought?

File: R/pkg/R/functions.R [hunk 9]
```diff
@@ -3948,26 +3869,18 @@ setMethod("grouping_bit",
...
-#' Returns the level of grouping.
-#'
+#' @section Details:
+#' \code{grouping_id}: Returns the level of grouping.
 #' Equals to \code{
 #' grouping_bit(c1) * 2^(n - 1) + grouping_bit(c2) * 2^(n - 2)  + ... + grouping_bit(cn)
 #' }
 #'
-#' @param x Column to compute on
-#' @param ... additional Column(s) (optional).
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: ditto here

File: R/pkg/R/generics.R (hunk 3 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: hmm, this is in `var` but also in `sd` too - perhaps should move to column_aggregate_functions

in fact, in your screen shot they are both listed https://github.com/apache/spark/pull/18025#issuecomment-303877945
`y,na.rm,use` and then `na.rm`

File: R/pkg/R/generics.R [hunk 10]
```diff
@@ -919,10 +920,9 @@ setGeneric("array_contains", function(x, value) { standardGeneric("array_contain
 #' @export
 setGeneric("ascii", function(x) { standardGeneric("ascii") })
 
-#' @param x Column to compute on or a GroupedData object.
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: I think we need to add this back to GroupedData, since that is now in a different rd?

**Review by felixcheung — COMMENTED** [review]
did another pass, thanks!

File: R/pkg/R/functions.R (hunk 6 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Yes, changed. Thanks for the suggestion.

File: R/pkg/R/functions.R (hunk 7 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: Yes, this newline is needed to separate blocks of examples.

File: R/pkg/R/functions.R (hunk 8 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: I agree it is less clear, but the impact is very minor if we have examples to illustrate passing additional columns. I now updated the doc of the argument as
`#' @param ... additional argument(s). For example, it could be used to pass additional Columns. `
And update the example to have multiple columns:
`head(select(df, countDistinct(df$gear, df$cyl)))`
Do the above changes address your concern?

File: R/pkg/R/generics.R (hunk 3 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: Good point. Moved to `column_aggregate_functions`.

File: R/pkg/R/generics.R (hunk 10 above)
Comment (actuaryzhang) [inline R/pkg/R/generics.R]: In this case, we will have to document `avg` on its own, like `count`, `first` and `last`. I cannot document the `x` param here since it will show up in the  doc for the column class. Interestingly, there is not even a doc of the `avg` method from the `GroupedData` class....

**Comment by actuaryzhang** [comment]
@felixcheung Your comments are all addressed now. Please let me know if there is anything else needed.

File: R/pkg/R/generics.R (hunk 10 above)
Comment (felixcheung) [inline R/pkg/R/generics.R]: yes, that's one of the code-gen methods that don't actually have documentation (which is a problem) but somehow inherit one from base:: that CRAN check doesn't complain about it

**Comment by actuaryzhang** [comment]
@felixcheung Could you take another look and let me know if there is anything else needed?

**Comment by HyukjinKwon** [comment]
@actuaryzhang, Are `corr`, `covar_pop` and `covar_samp` the same instances but missed? It looks these are also aggregate functions in `functions.scala` but these look missed here and I can see R has those functions.

**Comment by actuaryzhang** [comment]
@HyukjinKwon Thanks for catching this. They were incorrectly labeled as math functions instead of aggregate functions in SparkR. And that's why I did not change them. 
New commit fixed this now. Note they are still documented in their own Rd because there is also a method defined for SparkDataFrame. I made some cleaning and updated the example to be runnable.

File: R/pkg/R/functions.R (hunk 8 above)
Comment (felixcheung) [inline R/pkg/R/functions.R]: ok

File: R/pkg/R/functions.R [hunk 11]
```diff
@@ -361,10 +361,13 @@ setMethod("column",
...
 #' @rdname corr
 #' @name corr
-#' @family math functions
+#' @family aggregate functions
 #' @export
 #' @aliases corr,Column-method
-#' @examples \dontrun{corr(df$c, df$d)}
+#' @examples
+#' \dontrun{
+#' df <- createDataFrame(cbind(model = rownames(mtcars), mtcars))
```
Comment (felixcheung) [inline R/pkg/R/functions.R]: do we need space/newline in front of this example like the other ones?

File: R/pkg/R/stats.R [hunk 12]
```diff
@@ -52,22 +52,17 @@ setMethod("crosstab",
...
 #' @aliases cov,SparkDataFrame-method
 #' @family stat functions
 #' @export
 #' @examples
-#'\dontrun{
-#' df <- read.json("/path/to/file.json")
-#' cov <- cov(df, "title", "gender")
-#' }
+#'
+#' \dontrun{
```
Comment (felixcheung) [inline R/pkg/R/stats.R]: shouldn't the newline be after the dontrun?

File: R/pkg/R/stats.R [hunk 13]
```diff
@@ -52,22 +52,17 @@ setMethod("crosstab",
             collect(dataFrame(sct))
           })
 
-#' Calculate the sample covariance of two numerical columns of a SparkDataFrame.
```
Comment (felixcheung) [inline R/pkg/R/stats.R]: hmm, this is one of the tricky ones where there is one page for DataFrame & Columns.
I think it's useful to touch of how this works with a SparkDataFrame and keep this line in some form?

File: R/pkg/R/stats.R [hunk 14]
```diff
@@ -52,22 +52,17 @@ setMethod("crosstab",
...
-#'
 #' @param colName1 the name of the first column
 #' @param colName2 the name of the second column
-#' @return The covariance of the two columns.
```
Comment (felixcheung) [inline R/pkg/R/stats.R]: what would the `@return` line in the final doc?

File: R/pkg/R/functions.R (hunk 11 above)
Comment (actuaryzhang) [inline R/pkg/R/functions.R]: this one does not need the extra newline since it's in its own Rd and there are no examples before it.

File: R/pkg/R/stats.R (hunk 13 above)
Comment (actuaryzhang) [inline R/pkg/R/stats.R]: The method for SparkDataFrame is still there. I'm just removing redundant doc here.

File: R/pkg/R/stats.R (hunk 12 above)
Comment (actuaryzhang) [inline R/pkg/R/stats.R]: No. The newline should be between `@example` and `\dontrun` to separate multiple `dontruns`. 
image

File: R/pkg/R/stats.R (hunk 14 above)
Comment (actuaryzhang) [inline R/pkg/R/stats.R]: OK. I added this back. The doc should be very clear even without this return value. Indeed,  most functions do not document return value in SparkR. See what it looks like in the image attached in the next comment.

**Comment by actuaryzhang** [comment]
This is how the doc for column_aggregate_functions looks like (only snapshot of the main parts):

image
image
image
image

File: R/pkg/R/functions.R (hunk 11 above)
Comment (felixcheung) [inline R/pkg/R/functions.R]: great - I know we talk about it, but we might consider getting all examples on the Rd into one `\dontrun` block again. as of now it's very hard to review new PR without knowing whether a newline is needed or not...

Comment (felixcheung) [inline R/pkg/R/functions.R]: or maybe we should have a newline at the end of every `@example` block (when there are multiple examples on one Rd)? This way we don't have to know where goes first

File: R/pkg/R/stats.R (hunk 14 above)
Comment (felixcheung) [inline R/pkg/R/stats.R]: Possibly, but better clarity wouldn't hurt, right?

File: R/pkg/R/stats.R (hunk 13 above)
Comment (felixcheung) [inline R/pkg/R/stats.R]: I see. in that case can we add this in the `@details` for `cov`
I feel like this has bits of info that could be useful `Calculate the sample covariance of two numerical columns of a SparkDataFrame` - say, *numerical* columns of *one* SparkDataFrame, as supposed to `cov(df, df$name, df2$bar)`

**Comment by felixcheung** [comment]
very cool, thanks, I guess there's only this last comment https://github.com/apache/spark/pull/18025#discussion_r122631306

**Comment by actuaryzhang** [comment]
OK. Updated the doc for the cov method for SparkDataFrame.

**Comment by felixcheung** [comment]
haha. I like the `\emph`

**Comment by felixcheung** [comment]
AppVeyor failure is unfortunate. but it passed before a doc only change.

**Comment by felixcheung** [comment]
merged to master, thanks!
